electrumpersonalserver/server/peertopeer.py

```
@@ -0,0 +1,358 @@
+        self.uploaded_tx = False
+        self.connected = True
+
+    def on_heartbeat(self, p2p):
+        log.debug('broadcaster heartbeat')
+        VERACK_TIMEOUT = 40
+        GETDATA_TIMEOUT = 60
+        if not self.connected:
+            if (datetime.now() - self.time_marker).total_seconds() < VERACK_TIMEOUT:
+                return
+            log.debug('timed out of waiting for verack')
+        else:
+            if (datetime.now() - self.time_marker).total_seconds() < GETDATA_TIMEOUT:
+                return
+            log.debug('timed out of waiting for getdata, node already has tx')
+            self.uploaded_tx = True
+        p2p.close()
+
+    def handle_message(self, p2p, command, length, payload):
+        P2PMessageHandler.handle_message(self, p2p, command, length, payload)
+        ptr = [0]
+        if command == b'getdata':
+            count = read_var_int(ptr, payload)
+            for _ in range(count):
+                ptr[0] += 4
+                hash_id = payload[ptr[0] : ptr[0] + 32]
+                ptr[0] += 32
+                log.debug('hashid=' + hash_id.hex())
+                if hash_id == self.txid:
+                    log.debug('uploading tx')
+                    p2p.sock.sendall(p2p.create_message('tx', self.txhex))
+                    self.uploaded_tx = True
+                    p2p.close()
+
+def tor_broadcast_tx(txhex, tor_hostport, network, rpc):
+    ATTEMPTS = 8 # how many times to search for a node that accepts txes
+    try:
+        node_addrs = rpc.call("getnodeaddresses", [ATTEMPTS])
+        for i in range(ATTEMPTS):
+            remote_hostport = (node_addrs[i]["address"], node_addrs[i]["port"])
+            p2p_msg_handler = P2PBroadcastTx(txhex)
+            p2p = P2PProtocol(p2p_msg_handler, remote_hostport=remote_hostport,
+                network=network, socks5_hostport=tor_hostport, heartbeat_interval=20)
+            try:
+                p2p.run()
+            except:
```

Don't do this thing of swallowing errors. At least print a stack trace. Swallowing errors makes it very difficult to debug later. Ideally don't catch any-possible error but only the errors you need to. If you needed to place this try/except here because it was crashing then fix the crash in the lower level functions.